### PR TITLE
chore: Added client anticipation documentation

### PR DESCRIPTION
**Pull request by ShadauxCat**
#### **WARNING**: Please make your PR against the `develop` branch.
- - -
**Select the type of change:**

- [ ] Small Changes - Typos, formatting, slight revisions
- [x] New Content - New features, sections, pages, tutorials
- [ ] Site and Tools - Updates, maintenance, and new packages for the site and Docusaurus

**Purpose of the Pull Request:**
Adds documentation for the new Client Anticipation feature

**Issue Number:**

### CODE REVIEW

File: docs/basics/networkvariable.md
Comment (SamuelBellomo) [inline docs/basics/networkvariable.md:374]: This definition of extrapolation is different from the one used in the lag page (and also different from the one used in Netcode for Entities). There we use "extrapolation" as an "unclamped interpolation" where here you include lag to predict forward

Comment (SamuelBellomo) [reply docs/basics/networkvariable.md:374]: This is one of the reasons why single source of truth for concepts is important, so that we're at least consistent in our own terms, even if the industry isn't consistent.

Comment (ShadauxCat) [reply docs/basics/networkvariable.md:374]: I think both of those are extrapolation. The former is "linear extrapolation", which I have never personally thought was very useful since AFAIK it almost always gets wrong results. This is a more informed extrapolation taken from knowledge of how specific actors will behave.

The implementation supports both. Maybe I should add an example for linear extrapolation as well. Or do you think I should split this into two separate sections and come up with a different name for this approach (like maybe "forward simulation")?

Comment (SamuelBellomo) [reply docs/basics/networkvariable.md:374]: The difference is taking into account lag. Extrapolation is just a "continued" interpolation when you haven't received state from the server yet. In your case, your iterating forward taking into account lag. The timelines are different. Forward simulation sounds like a good name for it, not sure I'd touch "extrapolation". Maybe "lag based dead reckoning"?

Comment (ShadauxCat) [reply docs/basics/networkvariable.md:374]: Ah, gotcha... ok, yeah, that's different from what I had heard extrapolation defined as. (I had heard of it as "trying to predict where another player would be in a shooter to account for lag so the player doesn't miss shots they think they should have hit" and even the document I had read said "this is not very useful because it's usually wrong".) In the sense of making up for data arriving late from the server, I didn't include any support for that kind of thing here; interpolation keeps the position somewhat in the past but if the interpolation window is shorter than the time between updates it'll start to get hitchy.

Funny enough, though, when I looked up the term "dead reckoning" for networking, the first result said this:

> The basic notion of dead reckoning is agreement in advance on a set of algorithms that can be used by all player nodes to **extrapolate** the behavior of entities in the game

The industry's netcode terminology really is a mess, lol.

Dead reckoning sounds like the correct term... but part of me is inclined to say "forward simulation" because I feel like less experienced developers will understand that concept better, and because it seems less likely for someone with a different idea of what dead reckoning is to object to it.

Comment (SamuelBellomo) [reply docs/basics/networkvariable.md:374]: forward simulation sounds good yeah.
just FYI, I have this PR clarifying the N4E extrapolation [here](https://github.cds.internal.unity3d.com/unity/dots/pull/12104/files#diff-3e668504b65538739e92ee4795c6cbf3cf1b137616074628dec8e3bb476a9b71R4)

**Comment by jabbacakes** [comment]
Just a heads up that I've clocked this PR and will have a look at it ASAP once I'm done with Sam's in the N4E docs. 👍

**Comment by unity-cla-assistant** [comment]
[CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.multiplayer.docs?pullRequest=1215) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.multiplayer.docs?pullRequest=1215) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.cds.internal.unity3d.com/check/Unity-Technologies/com.unity.multiplayer.docs?pullRequest=1215) it.</sub>

**Comment by jabbacakes** [comment]
@ShadauxCat Heya, just doing a tidy up of PRs on the repo - is this one still relevant/active?

**Comment by ShadauxCat** [comment]
@jabbacakes Yes, this is still relevant, and needs to get out for 1.9. I'm waiting on some diagrams from @jilfranco-unity to finish it up.

**Comment by ShadauxCat** [comment]
@jabbacakes This PR is now fully ready for review.

File: docs/advanced-topics/client-anticipation.md
Comment (SamuelBellomo) [inline docs/advanced-topics/client-anticipation.md:1]: Little nitpick about the images. Not a blocker, but could make them a bit clearer. I'd try to put on top the "main" lesson" and have the other "support" client at the bottom. So for example, if the lesson is about reanticipation, I'd put the client timeline doing the reanticipation on top. Makes it easy to jump at the right spot when reading.
Not crazy important, as a reader I can figure it out, it just requires a bit more concentration.

File: static/img/sequence_diagrams/Anticipation/ServerAuthoritativeMultiClient.png
Comment (SamuelBellomo) [inline static/img/sequence_diagrams/Anticipation/ServerAuthoritativeMultiClient.png:1]: I'd add a "still green" comment bubble when Client 2 is setting to red, just to be consistent with Client 1

Comment (SamuelBellomo) [reply static/img/sequence_diagrams/Anticipation/ServerAuthoritativeMultiClient.png:1]: I'd also add in the title "no anticipation" just to make it clear this is with vanilla netvars

File: static/img/sequence_diagrams/Anticipation/StaleDataNoPolicy.png
Comment (SamuelBellomo) [inline static/img/sequence_diagrams/Anticipation/StaleDataNoPolicy.png:1]: For all the other diagrams, I'd maybe try to specify in the title that this is with anticipated netvars? Just to make it easy to understand which case is which

Comment (SamuelBellomo) [reply static/img/sequence_diagrams/Anticipation/StaleDataNoPolicy.png:1]: In the sense: anticipated vs vanilla

Comment (SamuelBellomo) [reply static/img/sequence_diagrams/Anticipation/StaleDataNoPolicy.png:1]: On another note, I'm a bit confused by the comment about flickering for Client 2. Cause that's not what the timeline is showing, it's still red all the time. I thought with no policy, It'd be green, red, blue, red? With this, Client 2 shows green, red, red, red.

File: docs/advanced-topics/client-anticipation.md
Comment (SamuelBellomo) [inline docs/advanced-topics/client-anticipation.md:43]:
```suggestion
One problem that an anticipation system needs to be able to solve is the problem of **stale data**. Stale data refers to updates from the server that represent actions that happened before your last request, and are actually going to be overwritten by that request.
```

File: static/img/sequence_diagrams/Anticipation/StaleDataNoPolicy.png
Comment (ShadauxCat) [reply static/img/sequence_diagrams/Anticipation/StaleDataNoPolicy.png:1]: Thanks for the callout on this. I hadn't noticed that. With no policy, it would NOT remain red, it would turn blue there.

The only diagram that uses vanilla network variables is the one labeled "server-authoritative". The others (including this one) are all using anticipated network variables.

File: docs/advanced-topics/client-anticipation.md
Comment (SamuelBellomo) [inline docs/advanced-topics/client-anticipation.md:61]: Part of this feels like implementation details that could be put a special comment box? Just to make it easier to skim over when fast reading?

File: docs/advanced-topics/client-anticipation.md
Comment (SamuelBellomo) [inline docs/advanced-topics/client-anticipation.md:63]: I'd list those two ways here before jumping in the details. This way users that are fast reading can just jump to it without having to read every details.

File: static/img/sequence_diagrams/Anticipation/StaleDataReanticipate.png
Comment (SamuelBellomo) [inline static/img/sequence_diagrams/Anticipation/StaleDataReanticipate.png:1]: For "client reanticipates inputs" I'd specify this is gameplay logic. Not NGO's client logic.

File: docs/advanced-topics/client-anticipation.md
Comment (SamuelBellomo) [inline docs/advanced-topics/client-anticipation.md:101]:
```suggestion
## Smoothing for Corrections
```

File: docs/advanced-topics/client-anticipation.md
Comment (SamuelBellomo) [inline docs/advanced-topics/client-anticipation.md:103]: I'd maybe try to add a link to this section at the top when explaining the basics of anticipation. I was reading the first paragraphs and I kept thinking "what about when there's desyncs?"

File: docs/advanced-topics/client-anticipation.md
Comment (SamuelBellomo) [inline docs/advanced-topics/client-anticipation.md:107]:
```suggestion
`Smooth()` takes a starting position and an ending position (which would usually be the previous anticipated value and either the new anticipated value or the current authoritative value) and a duration over which to perform the smoothing. For `AnticipatedNetworkVariable`, because the type is generic, it also requires a delegate to use to perform the smoothing operation (which takes a from value, a to value, and a percent value, and in many cases may just be `Mathf.Lerp`).
```

File: docs/advanced-topics/client-anticipation.md
Comment (SamuelBellomo) [inline docs/advanced-topics/client-anticipation.md:109]: Hopefully that's something Amy can help with, that doc has very long sentences :P

File: docs/advanced-topics/client-anticipation.md
Comment (SamuelBellomo) [inline docs/advanced-topics/client-anticipation.md:115]: Should that really be handled by Netcode? Couldn't users interpolate their own authoritative data or just use Physics interpolation for example?

File: docs/advanced-topics/client-anticipation.md
Comment (SamuelBellomo) [inline docs/advanced-topics/client-anticipation.md:117]: As a user, how would I access the authoritative value vs the visual value? What APIs do I use? For example, if my camera should track the interpolated value, how do I access it?

Comment (SamuelBellomo) [reply docs/advanced-topics/client-anticipation.md:117]: From going back above, I'm assuming that's gameObject.transform for interpolated? And this.GetComponent<AnticipatedNetworkTransform>().authoritativeState for authoritative value?

File: docs/learn/dealing-with-latency.md
Comment (SamuelBellomo) [inline docs/learn/dealing-with-latency.md:297]: > Because of the complexity inherent in building a full client prediction system, the details of that are left as an exercise for the user

Not sure I'd use "Because" here, there's not really a cause and effect.

Comment (SamuelBellomo) [reply docs/learn/dealing-with-latency.md:297]: I'd maybe try and list some of the complexity. As a sort of "here be dragons". Just so users don't waste time starting things they don't realize is more complex than they thought.

**Review by SamuelBellomo — APPROVED** [review]
Pre-approving once comments are addressed. Great doc! Much more straightforward now.

File: docs/advanced-topics/client-anticipation.md
Comment (NoelStephensUnity) [inline docs/advanced-topics/client-anticipation.md:74]: @jabbacakes
Not anything to hold this PR up, but we should see if the samples team has any bandwidth to create a series of companion examples to this documentation as it could be confusing to try to replicate the code required for this. Kitty already provided a generlized sample that has a lot of this done, so it might not be too time consuming to make something (based on the sample Kitty provided) that would walk users through this entire page of documentation.

File: docs/learn/dealing-with-latency.md
Comment (NoelStephensUnity) [reply docs/learn/dealing-with-latency.md:297]: +1
Even a "warning" callout box that implementing client prediction is a complex process and that while the client anticipation feature provides the framework to implement such a system the remaining code required is for advanced netcoders.

File: docs/learn/dealing-with-latency.md
Comment (NoelStephensUnity) [inline docs/learn/dealing-with-latency.md:360]: > with the most simple form simply being to set a value and let the server overwrite it later.

maybe:

with the simplest form of this being a client setting a value that the server overwrites later.

**Review by NoelStephensUnity — APPROVED** [review]
One minor adjustment to a sentence with some general comments on diagram adjustments and companion samples that can be done at a later time.

The only text copy adjustment I think should be adjusted is in the dealing-with-latency.md file.

Other than that, it all looks good!

**Review by jabbacakes — APPROVED** [review]
I was already two-thirds of the way through reviewing with suggestions when I got the 'it's okay to just push your changes' message, so... here are forty different suggestions ✨

Approving all the same, most of this is just copy editing to bring things in line with the style guide and shouldn't be controversial.

File: docs/advanced-topics/client-anticipation.md
Comment (NoelStephensUnity) [reply docs/advanced-topics/client-anticipation.md:66]: Added both approaches and created sub-header sections for each approach.

File: docs/advanced-topics/client-anticipation.md
Comment (ShadauxCat) [reply docs/advanced-topics/client-anticipation.md]: On the client side, yes, though it's such a common thing that I figured it would be useful to give users a method for it.

On the server side, smoothing is still used for interpolation, and it was a huge pain to implement it for AnticipatedNetworkTransform. They'd have a hard time implementing it themselves.

**Pull request merged**
